timeline_sync/models.py

```diff
@@ -0,0 +1,230 @@
+    actions = db.relationship('TimelineAction', lazy=False, uselist=True, cascade="all, delete-orphan")
+
+    data_source = db.Column(db.String(64), nullable=False)
+    source = db.Column(db.String(8), nullable=False)
+    create_time = db.Column(db.DateTime, nullable=False)
+    update_time = db.Column(db.DateTime, nullable=False)
+    topic_keys = db.Column(ARRAY(db.String), nullable=False)
+
+    @classmethod
+    def from_json(cls, pin_json, app_uuid, user_id, data_source, source, topic_keys):
+        try:
+            pin = cls(
+                guid=uuid.uuid4(),
+                id=pin_json['id'],
+                app_uuid=app_uuid,
+                user_id=user_id,
+                data_source=data_source,
+                source=source,
+                create_time=datetime.datetime.utcnow(),
+                topic_keys=topic_keys
+            )
+            pin.update_from_json(pin_json)
+            return pin
+        except (KeyError, ValueError):
+            return None
+
+    def update_from_json(self, pin_json):
+        self.time = valid_time(parse_time(pin_json['time']))
+        self.duration = pin_json.get('duration')
+        self.create_notification = TimelineNotification.from_json(pin_json.get('createNotification'), optional_time=True)
+        self.update_notification = TimelineNotification.from_json(pin_json.get('updateNotification'))
+        self.layout = TimelineLayout.from_json(pin_json['layout'])
+        self.reminders = TimelineReminder.from_json(pin_json.get('reminders'))
+        self.actions = TimelineAction.from_json(pin_json.get('actions'))
+        self.update_time = datetime.datetime.utcnow()
+
+    def to_json(self):
+        result = {
+            "time": time_to_str(self.time),
+            "layout": self.layout.to_json(),
+            "guid": self.guid,
+            "dataSource": self.data_source,
+            "source": self.source,
+            "createTime": time_to_str(self.create_time),
+            "updateTime": time_to_str(self.update_time),
+            "topicKeys": self.topic_keys
+        }
+
+        if self.duration is not None:
+            result["duration"] = self.duration
+        if self.create_notification is not None:
+            result["createNotification"] = self.create_notification.to_json()
+        if self.update_notification is not None:
+            result["updateNotification"] = self.update_notification.to_json()
+        if self.reminders is not None and len(self.reminders) > 0:
+            result["reminders"] = [reminder.to_json() for reminder in self.reminders]
+        if self.actions is not None and len(self.actions) > 0:
+            result["actions"] = [action.to_json() for action in self.actions]
+
+        return result
+
+
+db.Index('timeline_pin_appuuid_uid_pinid_index', TimelinePin.app_uuid, TimelinePin.user_id, TimelinePin.id, unique=True)
+
+
+class TimelineNotification(db.Model):
```

Comment:
I wonder whether it would make sense to store these notifications (and also reminders) as part of the `TimelinePin` in JSON[B] form. Most of our workload is going to be retrieving the various parts of a pin and transforming them to JSON — but we never need to do anything else with these. I suspect we'd save some IO/CPU time and some complexity if we just embedded these directly.

What do you think?